ethereum/sidecar/server.go

```diff
@@ -428,6 +537,342 @@ func (s *Server) startGRPCServer(
 	}
 }
 
+// observeAssetsUnlockedEvents monitors `AssetsUnlocked` events emitted on the
+// Mezo chain and stores unattested `AssetsUnlocked` events preparing them for
+// attestation. This routine consists of two parts:
+//   - initial check for unattested `AssetsUnlocked` events emitted when the sidecar
+//     was turned off
+//   - periodic check for new unattested `AssetsUnlocked` events
+func (s *Server) observeAssetsUnlockedEvents(ctx context.Context) error {
+	// At the start of the routine we need to learn which of the recent
+	// `AssetsUnlocked` events might require attestation in the `MezoBridge`
+	// smart contract. It's possible that the sidecar has been turned off
+	// for a significant amount of time while the Mezo chain was processing
+	// bridge-out requests.
+
+	// First fetch recent events. We consider such events as possibly unattested.
+	// The events are already sorted by their unlock sequence in ascending order.
+	recentEvents, err := s.fetchRecentAssetsUnlockedEvents(ctx)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to fetch recent AssetsUnlocked events: [%w]",
+			err,
+		)
+	}
+
+	unattestedEvents, err := s.findUnattestedAssetsUnlockedEvents(
+		ctx,
+		recentEvents,
+	)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to find unattested AssetsUnlocked events: [%w]",
+			err,
+		)
+	}
+
+	s.attestationMutex.Lock()
+	s.attestationQueue = unattestedEvents
+	s.attestationMutex.Unlock()
+
+	// Save the unlock sequence of the last event as the starting point for
+	// further event fetching.
+	s.lastAssetsUnlockedSequence = sdkmath.ZeroInt()
+	if len(recentEvents) > 0 {
+		s.lastAssetsUnlockedSequence = recentEvents[len(recentEvents)-1].UnlockSequence
```

Comment:
Double-checking my understanding: We are not caring about unconfirmed events here becauses we pushed them to the attestation queue. We do not re-check them later as this is the responsibility of the attestation loop. Is that correct?

Reply:
Yes.
In my opinion, the only job of the monitoring loop is to push unconfirmed events to the attestation queue.
Only the attestation loop should car about removing them once they are confirmed.